src/main/java/teammates/logic/api/EmailGenerator.java

```diff
@@ -1,2 +1,1083 @@
 // Reviewing this file for assignment
+package teammates.logic.api;
 
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import teammates.common.datatransfer.ErrorLogEntry;
+import teammates.common.datatransfer.attributes.AccountAttributes;
+import teammates.common.datatransfer.attributes.CourseAttributes;
+import teammates.common.datatransfer.attributes.DeadlineExtensionAttributes;
+import teammates.common.datatransfer.attributes.FeedbackSessionAttributes;
+import teammates.common.datatransfer.attributes.InstructorAttributes;
+import teammates.common.datatransfer.attributes.StudentAttributes;
+import teammates.common.util.Config;
+import teammates.common.util.Const;
+import teammates.common.util.EmailType;
+import teammates.common.util.EmailWrapper;
+import teammates.common.util.RequestTracer;
+import teammates.common.util.SanitizationHelper;
+import teammates.common.util.Templates;
+import teammates.common.util.Templates.EmailTemplates;
+import teammates.common.util.TimeHelper;
+import teammates.logic.core.CoursesLogic;
+import teammates.logic.core.FeedbackSessionsLogic;
+import teammates.logic.core.InstructorsLogic;
+import teammates.logic.core.StudentsLogic;
+
+/**
+ * Handles operations related to generating emails to be sent from provided templates.
+ *
+ * @see EmailTemplates
+ * @see EmailType
+ * @see EmailWrapper
+ */
+public final class EmailGenerator {
+    // feedback action strings
+    private static final String FEEDBACK_ACTION_SUBMIT_EDIT_OR_VIEW = "submit, edit or view";
+    private static final String FEEDBACK_ACTION_VIEW = "view";
+    private static final String FEEDBACK_ACTION_SUBMIT_OR_UPDATE =
+                            ", in case you have not submitted yet or wish to update your submission. ";
+    private static final String HTML_NO_ACTION_REQUIRED = "<mark>No action is required if you have already submitted</mark>";
```

Comment:
Consider not directly embedding HTML into the method itself, as this will make it difficult to actually modify and edit the email templates